agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/exporter/builders/AbstractTelemetryBuilder.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * ApplicationInsights-Java
+ * Copyright (c) Microsoft Corporation
+ * All rights reserved.
+ *
+ * MIT License
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of this
+ * software and associated documentation files (the ""Software""), to deal in the Software
+ * without restriction, including without limitation the rights to use, copy, modify, merge,
+ * publish, distribute, sublicense, and/or sell copies of the Software, and to permit
+ * persons to whom the Software is furnished to do so, subject to the following conditions:
+ * The above copyright notice and this permission notice shall be included in all copies or
+ * substantial portions of the Software.
+ * THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
+ * PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE
+ * FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+ * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+package com.microsoft.applicationinsights.agent.internal.exporter.builders;
+
+import static com.microsoft.applicationinsights.agent.internal.common.TelemetryTruncation.truncatePropertyValue;
+
+import com.microsoft.applicationinsights.agent.internal.common.Strings;
+import com.microsoft.applicationinsights.agent.internal.exporter.models.MonitorBase;
+import com.microsoft.applicationinsights.agent.internal.exporter.models.MonitorDomain;
+import com.microsoft.applicationinsights.agent.internal.exporter.models.TelemetryItem;
+import java.time.OffsetDateTime;
+import java.util.HashMap;
+import java.util.Map;
+
+public abstract class AbstractTelemetryBuilder {
+
+  private static final int MAX_PROPERTY_KEY_LENGTH = 150;
+  private static final int MAX_PROPERTY_VALUE_LENGTH = 8192;
+
+  protected static final int MAX_MEASUREMENT_KEY_LENGTH = 150;
+
+  private final TelemetryItem telemetryItem;
+
+  protected AbstractTelemetryBuilder(MonitorDomain data, String telemetryName, String baseType) {
+
+    telemetryItem = new TelemetryItem();
+    telemetryItem.setVersion(1);
+    telemetryItem.setName(telemetryName);
+
+    data.setVersion(2);
+
+    MonitorBase monitorBase = new MonitorBase();
+    telemetryItem.setData(monitorBase);
+    monitorBase.setBaseType(baseType);
+    monitorBase.setBaseData(data);
+  }
+
+  public void setTime(OffsetDateTime time) {
```

Comment:
should we match the name in basetelemetry? timestamp?  it's Date instead of OffsetDatetime in basetelemetry. do we need to be consistent?